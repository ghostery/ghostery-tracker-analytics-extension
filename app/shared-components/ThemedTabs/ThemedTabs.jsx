import React, { Component } from 'react';
import Tab from 'react-bootstrap/Tab';
import Nav from 'react-bootstrap/Nav';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';

import './ThemedTabs.scss';

class ThemedTabs extends Component {
  constructor(props) {
    super(props);

    this.state = { activeKey: this.setActiveKey() };
  }

  componentDidMount() {
    window.addEventListener('hashchange', this.handleHashChange, false);
  }

  componentWillUnmount() {
    window.removeEventListener('hashchange', this.handleHashChange, false);
  }

  handleHashChange = () => {
    const { passedSelectHandler } = this.props;
    const activeKey = this.setActiveKey();
    this.setState({ activeKey });
    if (!passedSelectHandler) { return; }
    passedSelectHandler(activeKey);
  }

  setActiveKey = () => {
    const hashParams = window.location.hash.split('tab=');
    if (hashParams.length === 0 || hashParams[0].length === 0) { return 1; }
    return hashParams[1][0];
  }

  tabSelectHandler = (selectedKey) => {
    const { hash } = window.location;
    window.location.hash = hash.replace(/(tab=.)/, `tab=${selectedKey}`);
  }

  _renderNavItem(tab, tabKey, tabClassNames, isLastOne) {
    const { activeKey } = this.state;

    const navLinkClassNames = ClassNames(
      'ThemedTabs__tabNavLink',
      { inLastTab: isLastOne },
    );

    const navLinkLabelClassNames = ClassNames(
      'ThemedTabs__tabLabel',
      {
        inActiveTab: tabKey === parseInt(activeKey, 10),
        inLastTab: isLastOne,
      },
    );

    return (
      <Nav.Item
        className={tabClassNames}
        key={tab.title}
      >
        <Nav.Link
          className={navLinkClassNames}
          eventKey={tabKey}
          active={tabKey === parseInt(activeKey, 10)}
        >
          <div className={navLinkLabelClassNames}>
            {tab.title}
          </div>
        </Nav.Link>
      </Nav.Item>
    );
  }

  _renderNavItems(tabs) {
    const tabClassNames = ClassNames(
      'ThemedTabs__tab',
      'd-flex',
      'justify-content-center',
      'align-items-center',
    );

    return (tabs.map((tab, i) => (
      this._renderNavItem(tab, i + 1, tabClassNames, i === tabs.length - 1)))
    );
  }

  render() {
    const { tabs, altTabs, stickyNavItems, mountOnEnter } = this.props;
    const { activeKey } = this.state;

    const tabsContainer = ClassNames({
      'ThemedTabs--reg': !altTabs,
      'ThemedTabs--alt': altTabs,
    });
    const tabsSelection = ClassNames('d-flex', {
      'justify-content-end': altTabs,
      themedTabsStickyPosition: stickyNavItems && activeKey === '2',
    });

    return (
      <div className={tabsContainer}>
        <Tab.Container
          className="ThemedTabs__tab"
          activeKey={activeKey}
          onSelect={this.tabSelectHandler}
          mountOnEnter={mountOnEnter}
          unmountOnExit
        >
          <div className={tabsSelection}>
            {!altTabs && <div className="ThemedTabs__whiteLine flex-grow-1" />}
            <Nav>
              {this._renderNavItems(tabs)}
            </Nav>
            {!altTabs && <div className="ThemedTabs__whiteLine flex-grow-1" />}
          </div>
          <Tab.Content>
            {tabs.map((tab, i) => (
              <Tab.Pane
                key={tab.title}
                eventKey={i + 1}
                active={`${i + 1}` === activeKey}
              >
                {React.cloneElement(tab.view, { active: `${i + 1}` === activeKey })}
              </Tab.Pane>
            ))}
          </Tab.Content>
        </Tab.Container>
      </div>
    );
  }
}

ThemedTabs.propTypes = {
  tabs: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.string.isRequired,
    view: PropTypes.element.isRequired,
  })).isRequired,
  altTabs: PropTypes.bool,
  stickyNavItems: PropTypes.bool,
  passedSelectHandler: PropTypes.func,
  mountOnEnter: PropTypes.bool,
};

ThemedTabs.defaultProps = {
  altTabs: false,
  stickyNavItems: false,
  passedSelectHandler: null,
  mountOnEnter: false,
};

export default ThemedTabs;
