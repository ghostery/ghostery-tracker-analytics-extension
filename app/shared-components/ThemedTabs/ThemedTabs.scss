@import '../../utils/scss/color-variables.scss';

.ThemedTabs--reg {
  margin-top: 25px;
}

.ThemedTabs--alt {
  margin-top: 15px;
  width: 100%;
}

.ThemedTabs__tabNavLink {
  box-sizing: border-box;
  border-top: 1px solid transparent;
  text-align: center;
  text-decoration: none;
  color: $white;

  &:focus {
    outline: 0;
  }
}

.ThemedTabs--reg .ThemedTabs__tabNavLink {
  border-bottom: 1px solid $white;
  padding: 0 1px;
  min-width: 170px;
  height: 26px;

  &.active {
    background-color: transparent;
    border: 1px solid $white;
    border-bottom: 1px solid transparent;
    padding: 0;
  }

  &.active.inLastTab {
    border-bottom: 1px solid $opaque-background-darker;
  }
}

.ThemedTabs--alt .ThemedTabs__tabNavLink {
  margin: 0 20px;
  background-color: transparent;
  border-bottom: 0;
  padding: 0;
  min-width: 90px;
  height: 23px;

  &.active {
    border: 0;
    border-bottom: 3px solid $ghosty-blue;
    font-weight: 600;
  }
}

.ThemedTabs--reg .ThemedTabs__tabLabel {
  background-color: $opaque-background-light;
  padding: 4px 0 3px;

  &.inActiveTab {
    background-color: transparent;
  }

  &.inActiveTab.inLastTab {
    background-color: $opaque-background-darker;
    border-bottom: 1px solid $opaque-background-darker;
  }
}

.ThemedTabs__whiteLine {
  position: relative;
  top: 25px;
  background-color: $white;
  height: 1px;
}

.themedTabsStickyPosition {
  position: sticky;
  top: 15px;
}
