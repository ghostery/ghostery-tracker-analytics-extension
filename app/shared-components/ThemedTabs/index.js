import ThemedTabs from './ThemedTabs.jsx';

export default ThemedTabs;
