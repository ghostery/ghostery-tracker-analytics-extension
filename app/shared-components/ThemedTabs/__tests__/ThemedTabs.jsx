import 'jsdom-global/register';
import React from 'react';
import { shallow, mount } from 'enzyme';
import path from 'path';
import fakeProps from 'react-fake-props';

import ThemedTabs from '../ThemedTabs';

describe('app/shared-components/ThemedTabs/', () => {
  describe('Snapshot tests with shallow', () => {
    test('ThemedTabs component should render', () => {
      const componentPath = path.join(__dirname, '../ThemedTabs.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<ThemedTabs {...props} />);
      expect(component).toMatchSnapshot();
    });
  });

  describe('functionality testing', () => {
    test('tabSelectHandler is called', () => {
      const componentPath = path.join(__dirname, '../ThemedTabs.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<ThemedTabs {...props} />);
      const instance = component.instance();
      const spy = jest.spyOn(instance, 'tabSelectHandler');
      instance.forceUpdate();

      component.find('ForwardRef.ThemedTabs__tab').simulate('select');
      expect(spy).toHaveBeenCalled();
    });
  });
});
