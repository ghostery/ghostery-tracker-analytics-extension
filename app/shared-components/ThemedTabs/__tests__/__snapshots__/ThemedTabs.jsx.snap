// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/shared-components/ThemedTabs/ Snapshot tests with shallow ThemedTabs component should render 1`] = `
<div
  className="ThemedTabs--reg"
>
  <ForwardRef
    activeKey={1}
    className="ThemedTabs__tab"
    mountOnEnter={false}
    onSelect={[Function]}
    unmountOnExit={true}
  >
    <div
      className="d-flex"
    >
      <div
        className="ThemedTabs__whiteLine flex-grow-1"
      />
      <Nav
        fill={false}
        justify={false}
      >
        <NavItem
          className="ThemedTabs__tab d-flex justify-content-center align-items-center"
          key="tabs.title"
        >
          <NavLink
            active={true}
            as={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "displayName": "SafeAnchor",
                "render": [Function],
              }
            }
            className="ThemedTabs__tabNavLink inLastTab"
            disabled={false}
            eventKey={1}
          >
            <div
              className="ThemedTabs__tabLabel inActiveTab inLastTab"
            >
              tabs.title
            </div>
          </NavLink>
        </NavItem>
      </Nav>
      <div
        className="ThemedTabs__whiteLine flex-grow-1"
      />
    </div>
    <Bootstrap(TabContent)>
      <TabPane
        active={false}
        eventKey={1}
        key="tabs.title"
      >
        <div
          active={false}
        >
          fake tabs.view element
        </div>
      </TabPane>
    </Bootstrap(TabContent)>
  </ForwardRef>
</div>
`;
